components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+
+use crate::{uniset::UnicodeSet, utils::deconstruct_range};
+
+/// UnicodeSet builder wrapper
+///
+/// Provides exposure to builder functions and conversion to UnicodeSet
+#[derive(Default)]
+pub struct UnicodeSetBuilder {
+    intervals: Vec<u32>,
+}
+
+impl UnicodeSetBuilder {
+    /// Returns empty UnicodeSetBuilder
+    pub fn new() -> UnicodeSetBuilder {
+        UnicodeSetBuilder { intervals: vec![] }
+    }
+
+    /// Returns a UnicodeSet and consumes the UnicodeSetBuilder
+    pub fn build(self) -> UnicodeSet {
+        UnicodeSet::from_inversion_list(self.intervals).unwrap()
+    }
+
+    /// Adds the range to the end of the builder
+    fn append(&mut self, start: u32, end: u32) {
+        self.intervals.push(start);
+        self.intervals.push(end);
+    }
+
+    /// Adds the range to the beginning of the builder in a single memory move
+    fn prepend(&mut self, start: u32, end: u32) {
+        let mut new = vec![start, end];
+        new.append(&mut self.intervals);
+        self.intervals = new;
+    }
+
+    /// Add the range to the UnicodeSetBuilder
+    ///
+    /// Accomplishes this through binary search for the start and end indices and merges intervals
+    /// in between with inplace memory. Performs `O(1)` operation if adding to end of list, and `O(N)` otherwise,
+    /// where `N` is the number of endpoints.
+    fn add(&mut self, start: u32, end: u32) {
+        if start >= end {
+            return;
+        }
+        if self.intervals.is_empty() {
+            self.append(start, end);
+        } else {
+            let last_ind = self.intervals.len() - 1;
+            if self.intervals[last_ind - 1] <= start {
+                if self.intervals[last_ind] < start {
+                    self.append(start, end);
+                } else {
+                    self.intervals[last_ind] = cmp::max(end, self.intervals[last_ind]);
+                }
+            } else if end <= self.intervals[1] {
+                if end < self.intervals[0] {
+                    self.prepend(start, end);
+                } else {
+                    self.intervals[0] = cmp::min(self.intervals[0], start);
+                }
+            } else {
+                let s = self.intervals.binary_search(&start);
+                let e = self.intervals.binary_search(&end);
+                let mut s_index = s.unwrap_or_else(|x| x);
+                let mut e_index = e.unwrap_or_else(|x| x);
+                let s_even: bool = s_index % 2 == 0;
+                let e_even: bool = e_index % 2 == 0;
+
+                if s_even && e_even && s.is_err() && e.is_err() && s_index == e_index {
+                    self.intervals.splice(s_index..e_index, vec![start, end]);
+                } else {
+                    let s_equals_e = s_index != e_index;
+                    if s_even {
+                        self.intervals[s_index] = start;
+                        s_index += 1;
+                    }
+                    if e_even && s_equals_e {
```

Comment:
If `e` is `Ok`, then you are adding a range whose endpoint is equal to the startpoint of another range.  You don't want to insert two adjacent numbers with the same values into the inversion list.  Please check.

Reply:
I add the explicit aligned tests (some of the test cases go over them but not explicitly/in the title) that you mentioned, and I still do not see this happening. Please advise on possible case, thanks!

Reply:
Start with: `10, 20, 30, 40`. Attempt to add: `15, 30`.

Variables:

```
s = Err(1)
e = Ok(2)
s_index = 1
e_index = 2
s_even = false
e_even = true
s_equals_e = false
```

You hit the condition `e_even && !s_equals_e`, and then promptly change `self.intervals[1]` to `30`.  Since `s_index` now equals `e_index`, you don't remove any elements.  You end up with `10, 30, 30, 40`, which violates the precondition of unique elements in the inversion list.

Reply:
I see the condition now, thanks!

I resolve by:
```rust
if e_even && !s_equals_e {
    if e.is_ok() {
        e_index += 1;
   }
   else {
        e_index -=1;
        self.intervals[e_index] = end;
    }
}
```

**Explanation**
In `[10, 20, 30, 40, 50, 60]`, the possible indices to hit the condition `if e_even && !s_equals_e` are `2 (30)` and `4 (50)` (`0 (10)` is handled as an edge case). 

When adding any interval where the binary search lands on the index:

`if e.is_ok()` (i.e. adding `[X, 30]`, `[X, 50]` where `X` is some arbitrary legal value)
Since the interval we are adding has every value between `X - 30`, while in our set we already have values between `30 - 40`, the value of `30` is redundant at index `2`, so we increase our `e_index` and drain the redundant vale

`if e.is_err()` then we handle as previously done by creating the new interval and draining the previous values